File: src/hbse/core/provider_tpm2.py
"""Linux TPM 2.0 provider backed by tpm2-tools.

This provider intentionally shells out only as an integration bridge. The
production spec still prefers direct TPM bindings to reduce process and file
handling exposure.
"""

from __future__ import annotations

import os
import hashlib
import shutil
import subprocess
import tempfile
from dataclasses import dataclass
from pathlib import Path

from hbse.core.keys import KEY_SIZE
from hbse.core.serialization import b64url_decode_no_padding, b64url_no_padding


TPM2_PROVIDER_ID = "linux-tpm2-tools-seal"


class TPM2ProviderError(RuntimeError):
    """Raised when the TPM provider cannot complete an operation."""


@dataclass(frozen=True)
class TPM2ProviderStatus:
    available: bool
    device_path: str
    tools_available: bool
    device_accessible: bool
    detail: str


@dataclass(frozen=True)
class LinuxTPM2ToolsProvider:
    device_path: str = "/dev/tpmrm0"

    def detect(self) -> TPM2ProviderStatus:
        tools = all(
            shutil.which(tool)
            for tool in ("tpm2_createprimary", "tpm2_create", "tpm2_load", "tpm2_unseal")
        )
        device = Path(self.device_path)
        accessible = device.exists() and os.access(device, os.R_OK | os.W_OK)
        if not device.exists():
            detail = f"{self.device_path} does not exist"
        elif not accessible:
            detail = f"{self.device_path} exists but is not readable/writable by this user"
        elif not tools:
            detail = "required tpm2-tools commands are not installed"
        else:
            detail = "TPM device and tpm2-tools are available"
        return TPM2ProviderStatus(
            available=bool(tools and accessible),
            device_path=self.device_path,
            tools_available=bool(tools),
            device_accessible=bool(accessible),
            detail=detail,
        )

    def wrap_root_key(self, *, vault_id: str, root_key: bytes) -> dict[str, object]:
        if len(root_key) != KEY_SIZE:
            raise ValueError("vault root key must be 32 bytes")
        self._require_available()
        with tempfile.TemporaryDirectory(prefix="hbse-tpm2-") as tmpdir:
            tmp = Path(tmpdir)
            secret_path = tmp / "root.key"
            primary_ctx = tmp / "primary.ctx"
            sealed_pub = tmp / "sealed.pub"
            sealed_priv = tmp / "sealed.priv"
            sealed_ctx = tmp / "sealed.ctx"
            secret_path.write_bytes(root_key)
            try:
                self._run("tpm2_createprimary", "-C", "o", "-G", "ecc", "-c", str(primary_ctx))
                self._run(
                    "tpm2_create",
                    "-C",
                    str(primary_ctx),
                    "-i",
                    str(secret_path),
                    "-u",
                    str(sealed_pub),
                    "-r",
                    str(sealed_priv),
                )
                self._run(
                    "tpm2_load",
                    "-C",
                    str(primary_ctx),
                    "-u",
                    str(sealed_pub),
                    "-r",
                    str(sealed_priv),
                    "-c",
                    str(sealed_ctx),
                )
                public_info = self._run("tpm2_readpublic", "-c", str(sealed_ctx), capture=True)
            finally:
                secret_path.write_bytes(b"\x00" * KEY_SIZE)
            return {
                "provider_id": TPM2_PROVIDER_ID,
                "vault_id": vault_id,
                "device_path": self.device_path,
                "parent_hierarchy": "owner",
                "public": b64url_no_padding(sealed_pub.read_bytes()),
                "private": b64url_no_padding(sealed_priv.read_bytes()),
                "public_info_sha256": b64url_no_padding(hashlib.sha256(public_info).digest()),
                "assurance_level": "A2",
                "warning": "tpm2-tools bridge provider; direct TPM bindings are preferred for production.",
            }

    def unwrap_root_key(self, *, binding: dict[str, object]) -> bytes:
        if binding.get("provider_id") != TPM2_PROVIDER_ID:
            raise TPM2ProviderError("unsupported TPM provider binding")
        self._require_available()
        with tempfile.TemporaryDirectory(prefix="hbse-tpm2-") as tmpdir:
            tmp = Path(tmpdir)
            primary_ctx = tmp / "primary.ctx"
            sealed_pub = tmp / "sealed.pub"
            sealed_priv = tmp / "sealed.priv"
            sealed_ctx = tmp / "sealed.ctx"
            sealed_pub.write_bytes(b64url_decode_no_padding(str(binding["public"])))
            sealed_priv.write_bytes(b64url_decode_no_padding(str(binding["private"])))
            self._run("tpm2_createprimary", "-C", "o", "-G", "ecc", "-c", str(primary_ctx))
            self._run(
                "tpm2_load",
                "-C",
                str(primary_ctx),
                "-u",
                str(sealed_pub),
                "-r",
                str(sealed_priv),
                "-c",
                str(sealed_ctx),
            )
            public_info = self._run("tpm2_readpublic", "-c", str(sealed_ctx), capture=True)
            expected_public_hash = str(binding.get("public_info_sha256", ""))
            actual_public_hash = b64url_no_padding(hashlib.sha256(public_info).digest())
            if expected_public_hash and expected_public_hash != actual_public_hash:
                raise TPM2ProviderError("TPM sealed object identity mismatch")
            root_key = self._run("tpm2_unseal", "-c", str(sealed_ctx), capture=True)
        if len(root_key) != KEY_SIZE:
            raise TPM2ProviderError("TPM returned invalid root key length")
        return root_key

    def self_test(self) -> TPM2ProviderStatus:
        status = self.detect()
        if not status.available:
            return status
        root_key = os.urandom(KEY_SIZE)
        binding = self.wrap_root_key(vault_id="self-test", root_key=root_key)
        unwrapped = self.unwrap_root_key(binding=binding)
        if unwrapped != root_key:
            raise TPM2ProviderError("TPM seal/unseal self-test mismatch")
        return status

    def _require_available(self) -> None:
        status = self.detect()
        if not status.available:
            raise TPM2ProviderError(status.detail)

    def _run(self, *args: str, capture: bool = False) -> bytes:
        env = {**os.environ, "TPM2TOOLS_TCTI": f"device:{self.device_path}"}
        try:
            result = subprocess.run(
                args,
                env=env,
                check=True,
                stdout=subprocess.PIPE if capture else subprocess.DEVNULL,
                stderr=subprocess.PIPE,
            )
        except FileNotFoundError as exc:
            raise TPM2ProviderError(f"missing command: {args[0]}") from exc
        except subprocess.CalledProcessError as exc:
            stderr = exc.stderr.decode("utf-8", errors="replace").strip()
            raise TPM2ProviderError(f"{args[0]} failed: {stderr}") from exc
        return result.stdout if capture else b""

File: hbse/src/provider_tpm2.rs
use std::fs;
use std::os::unix::fs::PermissionsExt;
use std::path::Path;
use std::process::{Command, Stdio};

use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use tempfile::tempdir;
use thiserror::Error;

use crate::keys::KEY_SIZE;
use crate::serialization::{b64url_decode_no_padding, b64url_no_padding};

pub const TPM2_PROVIDER_ID: &str = "linux-tpm2-tools-seal";

#[derive(Debug, Error)]
pub enum Tpm2ProviderError {
    #[error("vault root key must be 32 bytes")]
    InvalidRootKeyLength,
    #[error("unsupported TPM provider binding")]
    UnsupportedProvider,
    #[error("TPM provider unavailable: {0}")]
    Unavailable(String),
    #[error("missing command: {0}")]
    MissingCommand(String),
    #[error("{0} failed: {1}")]
    CommandFailed(String, String),
    #[error("TPM sealed object identity mismatch")]
    IdentityMismatch,
    #[error("TPM returned invalid root key length")]
    InvalidRootKeyLengthFromTpm,
    #[error("provider binding decode failed: {0}")]
    Decode(#[from] base64::DecodeError),
    #[error("io error: {0}")]
    Io(#[from] std::io::Error),
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Tpm2ProviderStatus {
    pub available: bool,
    pub device_path: String,
    pub tools_available: bool,
    pub device_accessible: bool,
    pub detail: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Tpm2ProviderBinding {
    pub provider_id: String,
    pub vault_id: String,
    pub device_path: String,
    pub parent_hierarchy: String,
    pub public: String,
    pub private: String,
    pub public_info_sha256: String,
    pub assurance_level: String,
    pub warning: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct LinuxTpm2ToolsProvider {
    pub device_path: String,
}

impl Default for LinuxTpm2ToolsProvider {
    fn default() -> Self {
        Self {
            device_path: "/dev/tpmrm0".to_string(),
        }
    }
}

impl LinuxTpm2ToolsProvider {
    pub fn new(device_path: impl Into<String>) -> Self {
        Self {
            device_path: device_path.into(),
        }
    }

    pub fn detect(&self) -> Tpm2ProviderStatus {
        let tools_available = [
            "tpm2_createprimary",
            "tpm2_create",
            "tpm2_load",
            "tpm2_unseal",
            "tpm2_readpublic",
        ]
        .into_iter()
        .all(command_exists);
        let device = Path::new(&self.device_path);
        let device_accessible = device.exists() && readable_writable(device);
        let detail = if !device.exists() {
            format!("{} does not exist", self.device_path)
        } else if !device_accessible {
            format!(
                "{} exists but is not readable/writable by this user",
                self.device_path
            )
        } else if !tools_available {
            "required tpm2-tools commands are not installed".to_string()
        } else {
            "TPM device and tpm2-tools are available".to_string()
        };
        Tpm2ProviderStatus {
            available: tools_available && device_accessible,
            device_path: self.device_path.clone(),
            tools_available,
            device_accessible,
            detail,
        }
    }

    pub fn wrap_root_key(
        &self,
        vault_id: &str,
        root_key: &[u8],
    ) -> Result<Tpm2ProviderBinding, Tpm2ProviderError> {
        if root_key.len() != KEY_SIZE {
            return Err(Tpm2ProviderError::InvalidRootKeyLength);
        }
        self.require_available()?;
        let dir = tempdir()?;
        let secret_path = dir.path().join("root.key");
        let primary_ctx = dir.path().join("primary.ctx");
        let sealed_pub = dir.path().join("sealed.pub");
        let sealed_priv = dir.path().join("sealed.priv");
        let sealed_ctx = dir.path().join("sealed.ctx");
        fs::write(&secret_path, root_key)?;
        let result = (|| {
            self.run(
                "tpm2_createprimary",
                &["-C", "o", "-G", "ecc", "-c", path_str(&primary_ctx)?],
                false,
            )?;
            self.run(
                "tpm2_create",
                &[
                    "-C",
                    path_str(&primary_ctx)?,
                    "-i",
                    path_str(&secret_path)?,
                    "-u",
                    path_str(&sealed_pub)?,
                    "-r",
                    path_str(&sealed_priv)?,
                ],
                false,
            )?;
            self.run(
                "tpm2_load",
                &[
                    "-C",
                    path_str(&primary_ctx)?,
                    "-u",
                    path_str(&sealed_pub)?,
                    "-r",
                    path_str(&sealed_priv)?,
                    "-c",
                    path_str(&sealed_ctx)?,
                ],
                false,
            )?;
            self.run("tpm2_readpublic", &["-c", path_str(&sealed_ctx)?], true)
        })();
        let _ = fs::write(&secret_path, vec![0u8; KEY_SIZE]);
        let public_info = result?;
        Ok(Tpm2ProviderBinding {
            provider_id: TPM2_PROVIDER_ID.to_string(),
            vault_id: vault_id.to_string(),
            device_path: self.device_path.clone(),
            parent_hierarchy: "owner".to_string(),
            public: b64url_no_padding(&fs::read(sealed_pub)?),
            private: b64url_no_padding(&fs::read(sealed_priv)?),
            public_info_sha256: b64url_no_padding(&Sha256::digest(public_info)),
            assurance_level: "A2".to_string(),
            warning:
                "tpm2-tools bridge provider; direct TPM bindings are preferred for production."
                    .to_string(),
        })
    }

    pub fn unwrap_root_key(
        &self,
        binding: &Tpm2ProviderBinding,
    ) -> Result<[u8; KEY_SIZE], Tpm2ProviderError> {
        if binding.provider_id != TPM2_PROVIDER_ID {
            return Err(Tpm2ProviderError::UnsupportedProvider);
        }
        self.require_available()?;
        let dir = tempdir()?;
        let primary_ctx = dir.path().join("primary.ctx");
        let sealed_pub = dir.path().join("sealed.pub");
        let sealed_priv = dir.path().join("sealed.priv");
        let sealed_ctx = dir.path().join("sealed.ctx");
        fs::write(&sealed_pub, b64url_decode_no_padding(&binding.public)?)?;
        fs::write(&sealed_priv, b64url_decode_no_padding(&binding.private)?)?;
        self.run(
            "tpm2_createprimary",
            &["-C", "o", "-G", "ecc", "-c", path_str(&primary_ctx)?],
            false,
        )?;
        self.run(
            "tpm2_load",
            &[
                "-C",
                path_str(&primary_ctx)?,
                "-u",
                path_str(&sealed_pub)?,
                "-r",
                path_str(&sealed_priv)?,
                "-c",
                path_str(&sealed_ctx)?,
            ],
            false,
        )?;
        let public_info = self.run("tpm2_readpublic", &["-c", path_str(&sealed_ctx)?], true)?;
        let actual_hash = b64url_no_padding(&Sha256::digest(public_info));
        if !binding.public_info_sha256.is_empty() && binding.public_info_sha256 != actual_hash {
            return Err(Tpm2ProviderError::IdentityMismatch);
        }
        let root_key = self.run("tpm2_unseal", &["-c", path_str(&sealed_ctx)?], true)?;
        root_key
            .try_into()
            .map_err(|_| Tpm2ProviderError::InvalidRootKeyLengthFromTpm)
    }

    pub fn self_test(&self) -> Result<Tpm2ProviderStatus, Tpm2ProviderError> {
        let status = self.detect();
        if !status.available {
            return Ok(status);
        }
        let root_key: [u8; KEY_SIZE] = rand::random();
        let binding = self.wrap_root_key("self-test", &root_key)?;
        let unwrapped = self.unwrap_root_key(&binding)?;
        if unwrapped != root_key {
            return Err(Tpm2ProviderError::CommandFailed(
                "tpm2_unseal".to_string(),
                "TPM seal/unseal self-test mismatch".to_string(),
            ));
        }
        Ok(status)
    }

    fn require_available(&self) -> Result<(), Tpm2ProviderError> {
        let status = self.detect();
        if status.available {
            Ok(())
        } else {
            Err(Tpm2ProviderError::Unavailable(status.detail))
        }
    }

    fn run(
        &self,
        command: &str,
        args: &[&str],
        capture: bool,
    ) -> Result<Vec<u8>, Tpm2ProviderError> {
        let mut process = Command::new(command);
        process
            .args(args)
            .env("TPM2TOOLS_TCTI", format!("device:{}", self.device_path))
            .stderr(Stdio::piped());
        if capture {
            process.stdout(Stdio::piped());
        } else {
            process.stdout(Stdio::null());
        }
        let output = process.output().map_err(|err| match err.kind() {
            std::io::ErrorKind::NotFound => Tpm2ProviderError::MissingCommand(command.to_string()),
            _ => Tpm2ProviderError::Io(err),
        })?;
        if !output.status.success() {
            return Err(Tpm2ProviderError::CommandFailed(
                command.to_string(),
                String::from_utf8_lossy(&output.stderr).trim().to_string(),
            ));
        }
        Ok(if capture { output.stdout } else { Vec::new() })
    }
}

fn command_exists(command: &str) -> bool {
    std::env::var_os("PATH").is_some_and(|paths| {
        std::env::split_paths(&paths).any(|path| {
            let candidate = path.join(command);
            candidate.is_file()
        })
    })
}

fn readable_writable(path: &Path) -> bool {
    let Ok(metadata) = fs::metadata(path) else {
        return false;
    };
    let mode = metadata.permissions().mode();
    mode & 0o600 != 0
}

fn path_str(path: &Path) -> Result<&str, Tpm2ProviderError> {
    path.to_str()
        .ok_or_else(|| Tpm2ProviderError::Unavailable("temporary path is not UTF-8".to_string()))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn detect_reports_missing_device() {
        let status = LinuxTpm2ToolsProvider::new("/definitely/not/a/tpm").detect();
        assert!(!status.available);
        assert!(!status.device_accessible);
        assert!(status.detail.contains("does not exist"));
    }
}
